core-bundle/src/Resources/contao/library/Contao/Database/Statement.php

```diff
@@ -257,17 +258,26 @@ public function query($strQuery='', array $arrParams = array(), array $arrTypes
 			throw new \Exception('Empty query string');
 		}
 
+		$arrTypes += array_fill(0, \count($arrParams), null);
 		$arrParams = array_map(
-			static function ($varParam)
+			static function ($key, $varParam) use (&$arrTypes)
 			{
+				// Automatically set type to boolean when no type is defined, otherwise
+				// PDO will convert "false" to an empty string (see https://bugs.php.net/bug.php?id=57157)
+				if (null === $arrTypes[$key] && \is_bool($varParam))
+				{
+					$arrTypes[$key] = ParameterType::BOOLEAN;
+				}
+
 				if (\is_string($varParam) || \is_bool($varParam) || \is_float($varParam) || \is_int($varParam) || $varParam === null)
 				{
 					return $varParam;
 				}
 
 				return serialize($varParam);
 			},
-			$arrParams
+			array_keys($arrParams),
+			array_values($arrParams)
```

Comment:
If it is possible that `$arrParams` is not a `list` we have to use array_values() for the first parameter as well.

```suggestion
			array_keys(array_values($arrParams)),
			array_values($arrParams)
```

If not, array_values would not be needed at all:

```suggestion
			array_keys($arrParams),
			$arrParams
```

Reply:
Good point. Actually I think we would need to support both cases, don't we? i.e.

```php
->prepare("SELECT * FROM tl_content WHERE invisible = ?")
->execute([false])
```
```
->prepare("SELECT * FROM tl_content WHERE invisible = :invisible")
->execute(['invisbile' => false])
```

So your second suggestion is correct. I'll do some further adjustments.